### PR TITLE
[draft] AMP Phase 3: OBBT presolve integration

### PR DESCRIPTION
**Pull request by bernalde**
## What changed
Adds the scoped checklist for the OBBT presolve slice.

## Why
OBBT is independent of the formulation work and should be tracked separately so its effect on bound tightening and iteration count is measurable.

## Implementation plan
- wire `run_obbt` into AMP before the main loop
- expose the option through the public solver interface
- add tests for bound tightening and iteration-count improvement

## Validation
- Reviewed the staged tracking doc locally
- Branch is pushed and ready for follow-up commits

### CODE REVIEW

**Review by bernalde — COMMENTED** [review]
Overall this is a clean integration of OBBT as an AMP presolve step. The wiring through solver.py, the OBBT call placement before partition initialization, and the bound flow into `initialize_partitions` are all correct. Tests pass (72 passed, 2 known xfails). The cleanup changes (dead-code removal in partition_selection.py, unused imports in milp_relaxation.py, trilinear key construction in term_classifier.py) are all safe.

A few points to address below, mostly about time management and test robustness.

File: python/discopt/solvers/amp.py [hunk 1]
```diff
@@ -454,6 +451,26 @@ def _from_minimization_space(value: float) -> float:
         len(terms.general_nl),
     )
 
+    # Tighten the initial McCormick domain before selecting partition bounds.
+    if presolve_bt:
+        try:
+            from discopt._jax.obbt import run_obbt
+
+            obbt_result = run_obbt(
```
Comment (bernalde) [inline python/discopt/solvers/amp.py]: This call has no time budget. `run_obbt` accepts `time_limit_per_lp` but it is not wired here. If a model has many variables, OBBT will emit 2*n LP solves before the AMP loop even starts, potentially consuming most of the wall-clock budget.

Consider forwarding a fraction of the remaining time, e.g.:
```python
obbt_result = run_obbt(
    model,
    lb=flat_lb.copy(),
    ub=flat_ub.copy(),
    time_limit_per_lp=max(0.1, (time_limit * 0.1) / max(1, 2 * n_orig)),
)
```
The exact fraction is a tuning choice, but without any limit this is unbounded.

File: python/discopt/solvers/amp.py [hunk 2]
```diff
@@ -454,6 +451,26 @@ def _from_minimization_space(value: float) -> float:
...
+                ub=flat_ub.copy(),
+            )
+            if obbt_result.n_tightened > 0:
+                flat_lb = obbt_result.tightened_lb
+                flat_ub = obbt_result.tightened_ub
+                logger.info(
+                    "AMP: OBBT tightened %d bounds before partitioning",
+                    obbt_result.n_tightened,
+                )
+        except Exception as err:
```
Comment (bernalde) [inline python/discopt/solvers/amp.py]: Catching bare `Exception` here will silently swallow real bugs, e.g. a `TypeError` in `_extract_linear_constraints` or a numpy `MemoryError`. Since the only expected failures are `ImportError` (if obbt.py is missing) and `RuntimeError` / solver errors from HiGHS, narrowing the catch would make development-time bugs surface faster.

At a minimum, consider logging at `error` level so failures show up in default log output even if the solver continues.

File: python/discopt/solvers/amp.py [hunk 3]
```diff
@@ -454,6 +451,26 @@ def _from_minimization_space(value: float) -> float:
...
+            obbt_result = run_obbt(
+                model,
+                lb=flat_lb.copy(),
+                ub=flat_ub.copy(),
+            )
+            if obbt_result.n_tightened > 0:
+                flat_lb = obbt_result.tightened_lb
+                flat_ub = obbt_result.tightened_ub
+                logger.info(
+                    "AMP: OBBT tightened %d bounds before partitioning",
```
Comment (bernalde) [inline python/discopt/solvers/amp.py]: The log message only reports `n_tightened`. Consider also logging `obbt_result.total_lp_time` so the user can see whether OBBT was a net win or a time sink. Something like:
```python
logger.info(
    "AMP: OBBT tightened %d bounds in %.3fs before partitioning",
    obbt_result.n_tightened,
    obbt_result.total_lp_time,
)
```

File: python/tests/test_amp.py [hunk 4]
```diff
@@ -1223,6 +1234,91 @@ def fake_build(model, terms, disc_state, incumbent, oa_cuts=None):
...
+            iteration_callback=lambda info: iters_with.append(info["iteration"]),
+        )
+
+        assert result_without.status == "optimal"
+        assert result_with.status == "optimal"
+        assert result_without.objective is not None
+        assert result_with.objective is not None
+        assert abs(result_without.objective - 0.25) <= 1e-2
+        assert abs(result_with.objective - 0.25) <= 1e-2
+        assert len(iters_with) < len(iters_without)
```
Comment (bernalde) [inline python/tests/test_amp.py]: This strict inequality between iteration counts is fragile. AMP iteration counts depend on several solver internals (warm-start quality, HiGHS LP precision, NLP convergence path) that could shift with dependency updates. If a HiGHS or JAX version bump causes both runs to converge in the same number of iterations, this test will fail even though OBBT is working correctly.

A safer formulation would assert `<=` and separately verify that OBBT actually tightened bounds (which `test_obbt_presolve_tightens_bilinear_demo_bounds` already does). Or keep the strict `<` but add a comment explaining the expected margin and the relaxed `rel_gap=0.55` that makes this work.

File: python/tests/test_amp.py [hunk 5]
```diff
@@ -105,6 +99,16 @@ def _build_nlp3() -> Model:
     return m
 
 
+def _make_obbt_demo() -> Model:
```
Comment (bernalde) [inline python/tests/test_amp.py:102]: All OBBT tests use this model, which only has an equality constraint (`x + y == 1`). That exercises only the `A_eq / b_eq` path in `_extract_linear_constraints`. The `A_ub / b_ub` path (for `<=` and `>=` constraints) is never tested.

Consider adding a small variant, e.g. `x <= 1; y <= 1; x + y >= 0.5` alongside the equality, and verifying that OBBT tightens the upper bounds there too.

File: python/discopt/solvers/amp.py (hunk 1 above)
Comment (bernalde) [reply python/discopt/solvers/amp.py]: Fixed in [d7a636e](https://github.com/bernalde/discopt/commit/d7a636e0989302d98c3059aa764f780cd6cf629c). `solve_amp` now computes a bounded OBBT per-LP budget from the remaining wall-clock time with `_default_obbt_time_limit_per_lp(...)` and forwards it as `time_limit_per_lp` to `run_obbt`. The forwarding path is covered by the updated OBBT tests on this branch.

File: python/discopt/solvers/amp.py (hunk 2 above)
Comment (bernalde) [reply python/discopt/solvers/amp.py]: Fixed in [d7a636e](https://github.com/bernalde/discopt/commit/d7a636e0989302d98c3059aa764f780cd6cf629c). The bare `Exception` catch is gone; this now only swallows `ImportError` for an unavailable OBBT module. Runtime failures inside OBBT or the LP layer are no longer masked and will surface during development.

File: python/discopt/solvers/amp.py (hunk 3 above)
Comment (bernalde) [reply python/discopt/solvers/amp.py]: Fixed in [d7a636e](https://github.com/bernalde/discopt/commit/d7a636e0989302d98c3059aa764f780cd6cf629c). The OBBT log now reports both `n_tightened` and `obbt_result.total_lp_time`, so the presolve cost is visible alongside the tightening effect.

File: python/tests/test_amp.py (hunk 4 above)
Comment (bernalde) [reply python/tests/test_amp.py]: Fixed in [d7a636e](https://github.com/bernalde/discopt/commit/d7a636e0989302d98c3059aa764f780cd6cf629c). The iteration-count assertion is now `<=`, and the test comment makes the intent explicit: actual bound tightening is checked separately, while this one just guards against OBBT making the AMP loop worse. After rebasing on the merged Phase 2 branch, the spy in the same test also needed the `initialize_partitions(...)` kwarg update in [cf0f4e9](https://github.com/bernalde/discopt/commit/cf0f4e9523d2e63b706e3a9a078d66d5abe21820).

File: python/tests/test_amp.py (hunk 5 above)
Comment (bernalde) [reply python/tests/test_amp.py:102]: Fixed in [d7a636e](https://github.com/bernalde/discopt/commit/d7a636e0989302d98c3059aa764f780cd6cf629c). I added `_make_obbt_ineq_demo()` plus `test_obbt_presolve_tightens_inequality_demo_bounds`, so the OBBT tests now exercise the `A_ub / b_ub` extraction path instead of only the equality case.

**Comment by bernalde** [comment]
Addressed the Phase 3 review comments on the rebased branch head [cf0f4e9](https://github.com/bernalde/discopt/commit/cf0f4e9523d2e63b706e3a9a078d66d5abe21820). The main review fixes are in [d7a636e](https://github.com/bernalde/discopt/commit/d7a636e0989302d98c3059aa764f780cd6cf629c), and the branch is now restacked onto `feature/amp-completion/phase-1-robustness`.

Validation on `cf0f4e9`:
`~/.local/bin/uv run pytest python/tests/test_amp.py -m "not slow"` -> `83 passed, 1 deselected, 1 xfailed`
`bash scripts/ci_local.sh` -> `1930 passed, 75 skipped, 219 deselected, 1 xfailed`
